### PR TITLE
Improve configuration management

### LINKED ISSUE #108: RefBox consumes all available CPU power when started with plc connection mps
The default config now is configured to no longer have mockup connection mode configured for the mps.
If you just start the RefBox without any MPS stations to connect to, your CPU gets fried, this is probably avoidable.

### PR DESCRIPTION
**Pull request by TarikViehmann**
As @morxa commented in https://github.com/robocup-logistics/rcll-refbox/pull/102, it currently is rather annoying to constantly modify the configuration of the refbox.

Also, if the refbox is just installed as a binary, there is currently no great way of customizing it.

With this PR the situation should hopefully become better.

Firstly, the config is split into different components, which allows for fine-grained customization.

Secondly, the main executable parses input settings which allow to switch between different configuration files for each defined component.
Components are defined by simply creating folders `<folder-name>` within `cfg/` that contain a default file named `default_<folder-name>.yaml`
If the refbox is started without arguments, it loads all these default files of the components, with the supplied defaults in this PR, exactly the config values of the previous `config.yaml` are loaded. 
To switch out individual component, a command-line option simply loads an alternative file:
For example `--cfg-mps mps/mockup_mps.yaml` will load the mockup settings instead of the plc connection options defined in `default_mps.yaml`.

Lastly, in case you rather want to have the old behavior and load your own custom stuff, there are two options:
 `--no-default-cfg` to disable loading of the default components
 `--cfg-custom <yaml-file>` to load a configuration file after all other yaml files are loaded, hence you can be sure that the config defined here are overriding all possibly conflicting options loaded through components.

So invoking `llsf-refbox --no-default-cfg --cfg-custom <one config file containing everything>` would allow to use the refbox as before.

Also, while working on this PR i encountered issue #108, which i also fixed.
So this PR resolves #108.

### CODE REVIEW

File: src/libs/mps_comm/opcua/machine.cpp [hunk 1]
```diff
@@ -170,7 +170,9 @@ OpcUaMachine::connect()
 	if (connection_mode_ == MOCKUP) {
 		return;
 	}
-	while (!reconnect()) {}
+	while (!reconnect()) {
+		std::this_thread::sleep_for(std::chrono::milliseconds(40));
```
Comment (snoato) [inline src/libs/mps_comm/opcua/machine.cpp:174]: Maybe this should be configurable, a variable, or a macro as it is used multiple times.

**Review by snoato — CHANGES_REQUESTED** [review]
Great job improving the usability and ease of configuration of the refbox! Only one minor comment.

**Comment by morxa** [comment]
Is there a reason to store the generated config as file? I would have expected that you just load the files sequentially into the config and therefore override values with the config files loaded last

**Comment by TarikViehmann** [comment]
> Is there a reason to store the generated config as file? I would have expected that you just load the files sequentially into the config and therefore override values with the config files loaded last

Thats also what i thought, hence i first tried to simply call `config_->load()`on each file. Sadly, this wipes all the previously loaded configs. After your comment i revisited the yaml parser, turns out the fix for not overwriting on subsequent `load()` calls is easy. But, then i found another issue: All tools also want to parse the config, which means a refbox started with command-line options would require the tools to be configurable in the same manner, if we get rid of generated configuration files.

File: src/libs/mps_comm/opcua/machine.cpp (hunk 1 above)
Comment (TarikViehmann) [reply src/libs/mps_comm/opcua/machine.cpp]: i changed that to be a inline constexpr, good catch!

**Comment by morxa** [comment]
> > Is there a reason to store the generated config as file? I would have expected that you just load the files sequentially into the config and therefore override values with the config files loaded last
> 
> Thats also what i thought, hence i first tried to simply call `config_->load()`on each file. Sadly, this wipes all the previously loaded configs. After your comment i revisited the yaml parser, turns out the fix for not overwriting on subsequent `load()` calls is easy. But, then i found another issue: All tools also want to parse the config, which means a refbox started with command-line options would require the tools to be configurable in the same manner, if we get rid of generated configuration files.

You could use `include` and have several root configs that include the right config files, if you want to have multiple configurations.

AFAIK the refbox config parser is the same as in fawkes, so something like [this](https://github.com/fawkesrobotics/fawkes/blob/master/cfg/config.yaml) should just work.

**Comment by TarikViehmann** [comment]
> AFAIK the refbox config parser is the same as in fawkes, so something like [this](https://github.com/fawkesrobotics/fawkes/blob/master/cfg/config.yaml) should just work.

So that means having one root config file for each of the "standard" configs that we provide, right?
Yea that would be possible, and i initially thought about that (am still open to that idea). My argument against that was that
for each "standard" config we want to provide (e.g., for each challenge or for mockup vs plc  - which we need to easily configure if we switch between gazebo simulation and skiller simulation, right? -), we also need a root config, which might get out of hand.

**Comment by TarikViehmann** [comment]
We could also simply do the the following:
The refbox sequentially loads all the provided config files and does not generate a root config by default. Optionally, you can also dump the resulting config, which can be used if you want to use the tools.
What do you think?

**Review by snoato — DISMISSED** [review]

**Review by snoato — APPROVED** [review]

**Pull request merged**